File: front/src/utils/passwordValidator.js
/**
 * Validador de senha com regras de complexidade
 * Deve atender aos critérios:
 * - Mínimo 8 caracteres
 * - Pelo menos 1 maiúscula
 * - Pelo menos 1 minúscula
 * - Pelo menos 1 número
 * - Pelo menos 1 caractere especial
 */

export const validatePassword = (password) => {
  const errors = [];

  if (!password || password.trim() === '') {
    errors.push('A senha é obrigatória');
    return errors;
  }

  // Mínimo 8 caracteres
  if (password.length < 8) {
    errors.push('A senha deve ter pelo menos 8 caracteres');
  }

  // Pelo menos 1 maiúscula
  if (!/[A-Z]/.test(password)) {
    errors.push('A senha deve conter pelo menos uma letra maiúscula');
  }

  // Pelo menos 1 minúscula
  if (!/[a-z]/.test(password)) {
    errors.push('A senha deve conter pelo menos uma letra minúscula');
  }

  // Pelo menos 1 número
  if (!/[0-9]/.test(password)) {
    errors.push('A senha deve conter pelo menos um número');
  }

  // Pelo menos 1 caractere especial
  if (!/[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>?/~`]/.test(password)) {
    errors.push('A senha deve conter pelo menos um caractere especial (!@#$%^&*()_+-=[]{}|;\':"\\,.<>?/~`)');
  }

  return errors;
};

export const isValidPassword = (password) => {
  return validatePassword(password).length === 0;
};

export const getPasswordCriteria = () => {
  return [
    'Pelo menos 8 caracteres',
    'Pelo menos uma letra maiúscula (A-Z)',
    'Pelo menos uma letra minúscula (a-z)',
    'Pelo menos um número (0-9)',
    'Pelo menos um caractere especial (!@#$%^&*()_+-=[]{}|;\':"\\,.<>?/~`)'
  ];
};

export const getPasswordStrengthLevel = (password) => {
  if (!password) return 0;
  
  const criteria = [
    password.length >= 8,
    /[A-Z]/.test(password),
    /[a-z]/.test(password),
    /[0-9]/.test(password),
    /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>?/~`]/.test(password)
  ];
  
  return criteria.filter(Boolean).length;
};

export const getPasswordStrengthText = (level) => {
  switch (level) {
    case 0:
    case 1:
      return 'Muito fraca';
    case 2:
      return 'Fraca';
    case 3:
      return 'Média';
    case 4:
      return 'Forte';
    case 5:
      return 'Muito forte';
    default:
      return 'Inválida';
  }
};

export const getPasswordStrengthColor = (level) => {
  switch (level) {
    case 0:
    case 1:
      return '#e74c3c';
    case 2:
      return '#f39c12';
    case 3:
      return '#f1c40f';
    case 4:
      return '#27ae60';
    case 5:
      return '#2ecc71';
    default:
      return '#e74c3c';
  }
}; 

File: core/src/main/java/com/senac/dentalsync/core/util/PasswordValidator.java
package com.senac.dentalsync.core.util;

import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;

import java.util.regex.Pattern;

/**
 * Utilitário para validação de senhas com regras de complexidade
 */
public class PasswordValidator {

    // Padrões regex para validação
    private static final Pattern UPPERCASE_PATTERN = Pattern.compile(".*[A-Z].*");
    private static final Pattern LOWERCASE_PATTERN = Pattern.compile(".*[a-z].*");
    private static final Pattern DIGIT_PATTERN = Pattern.compile(".*[0-9].*");
    private static final Pattern SPECIAL_CHAR_PATTERN = Pattern.compile(".*[!@#$%^&*()_+\\-=\\[\\]{};':\"\\\\|,.<>?/~`].*");

    /**
     * Valida se a senha atende aos critérios de complexidade
     * 
     * @param password A senha a ser validada
     * @throws ResponseStatusException Se a senha não atender aos critérios
     */
    public static void validatePassword(String password) {
        if (password == null || password.trim().isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Senha é obrigatória");
        }

        // Mínimo 8 caracteres
        if (password.length() < 8) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, 
                "A senha deve ter pelo menos 8 caracteres");
        }

        // Pelo menos 1 letra maiúscula
        if (!UPPERCASE_PATTERN.matcher(password).matches()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, 
                "A senha deve conter pelo menos uma letra maiúscula");
        }

        // Pelo menos 1 letra minúscula
        if (!LOWERCASE_PATTERN.matcher(password).matches()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, 
                "A senha deve conter pelo menos uma letra minúscula");
        }

        // Pelo menos 1 número
        if (!DIGIT_PATTERN.matcher(password).matches()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, 
                "A senha deve conter pelo menos um número");
        }

        // Pelo menos 1 caractere especial
        if (!SPECIAL_CHAR_PATTERN.matcher(password).matches()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, 
                "A senha deve conter pelo menos um caractere especial (!@#$%^&*()_+-=[]{}|;':\"\\,.<>?/~`)");
        }
    }

    /**
     * Verifica se a senha atende aos critérios sem lançar exceção
     * 
     * @param password A senha a ser validada
     * @return true se a senha é válida, false caso contrário
     */
    public static boolean isValidPassword(String password) {
        try {
            validatePassword(password);
            return true;
        } catch (ResponseStatusException e) {
            return false;
        }
    }

    /**
     * Retorna uma mensagem com os critérios de senha
     * 
     * @return String com os critérios
     */
    public static String getPasswordCriteria() {
        return "A senha deve conter:\n" +
               "- Pelo menos 8 caracteres\n" +
               "- Pelo menos uma letra maiúscula (A-Z)\n" +
               "- Pelo menos uma letra minúscula (a-z)\n" +
               "- Pelo menos um número (0-9)\n" +
               "- Pelo menos um caractere especial (!@#$%^&*()_+-=[]{}|;':\"\\,.<>?/~`)";
    }
} 